### DISCUSSION TITLE
How do I properly use RichTextLayout.GetGlyphInfoByIndex to find the position of a character

### DISCUSSION THREAD

**Discussion by bittiez** [rds1983/FontStashSharp#80] [Q&A] [answered]
I'm trying to create an input field; I have everything working except displaying the caret and displaying a selection because I can't get character positions using `GetGlyphInfoByIndex`, it doesn't seem to matter what index I use it always throws an out of range error.

```
System.ArgumentOutOfRangeException
  HResult=0x80131502
  Message=Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
  Source=mscorlib
  StackTrace:
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at FontStashSharp.RichText.TextChunk.GetGlyphInfoByIndex(Int32 index)
   at FontStashSharp.RichText.TextLine.GetGlyphInfoByIndex(Int32 index)
   at FontStashSharp.RichText.RichTextLayout.GetGlyphInfoByIndex(Int32 charIndex)
   at ClassicUO.Game.UI.Gumps.ModernOptionsGump.InputField.StbTextBox.UpdateCaretScreenPosition()
```

That error was thrown using this method:
```csharp
                    if (!string.IsNullOrEmpty(_rendererText.Text))
                    {
                        var g = _rendererText.RTL.GetGlyphInfoByIndex(0);
                        if (g != null && g.HasValue)
                        {
                            _caretScreenPosition = new Point(g.Value.Bounds.X, g.Value.Bounds.Y);
                        }
                    }
```

Using 0 as the index, I'm not sure what I'm doing wrong here.

`_rendererText` is a wrapper around your RichTextLayout class, the `.RTL` is the instance of your RichTextLayout

**Comment by bittiez** [comment]
Now that i've had my coffee I realize the glyph would only return the size of that char, which I do need that info but I also need to find the coords of an index, or the width of all the characters before that index?

**Comment by rds1983** [comment]
Check code of Myra's TextBox: https://github.com/rds1983/Myra/blob/master/src/Myra/Graphics2D/UI/Simple/TextBox.cs
It solves exactly same task.

**Reply by bittiez** [reply]
TY I will look it over

**Reply by bittiez** [reply]
So it looks like Myra's is doing the exact thing I was doing, I rewrote it again and still I'm getting index out of range.

`System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index'`

My code:
```
                    if (!string.IsNullOrEmpty(_rendererText.Text))
                    {
                        if (Stb.CursorIndex < _rendererText.Text.Length)
                        {
                            var glyph = _rendererText.RTL.GetGlyphInfoByIndex(Stb.CursorIndex);
                            if(glyph != null)
                            {
                                _caretScreenPosition = new Point(glyph.Value.Bounds.X, glyph.Value.Bounds.Y);
                            }
                        }
                    }
```

And Myras: https://github.com/rds1983/Myra/blob/001a0bc630e1544f13d28f65220776b32347c610/src/Myra/Graphics2D/UI/Simple/TextBox.cs#L1323

I'm really not understanding how the index is out of range when it checks that before it even calls the `GetGlyphInfoByIndex`

**Comment by bittiez** [comment, marked as answer]
For possible future peoples I did figure out the issue, I didn't realize I needed to set `CalculateGlyphs = true` on the RichTextLayout 👍
